File: number theory and crypto/javascript/prime.js
// is prime
function isPrime(n) {
    if (n < 2) return false;
    else if (n == 2) return true;
    else if (n % 2 == 0) return false;
    for (let i = 3; i <= Math.sqrt(n); i += 2) {
        if (n % i == 0) return false;
    }
    return true;
}
function sieve(n) {
    // declare array upto n 
    let arr = new Array(n + 1).fill(true);
    // 0 and 1 are not prime
    arr[0] = arr[1] = false;
    // iterate upto sqrt(n)
    for (let i = 2; i <= Math.sqrt(n); i++) {
        // if i is prime
        if (arr[i]) {
            // mark all multiples of i as false
            for (let j = i * i; j <= n; j += i) {
                arr[j] = false;
            }
        }
    }
    // return array of prime numbers
    return arr;
}
// test
console.log(isPrime(49));

let n = 100;
let primes = sieve(n);
for (let i = 0; i <= n; i++) {
    if (primes[i]) console.log(i);
}

function test() {
    // we will use sieve to generate prime numbers upto 100 and then check it against isPrime function
    let n = 100;
    let primes = sieve(n);
    for (let i = 0; i <= n; i++) {
        if (primes[i] != isPrime(i)) {
            // raise exception
            throw new Error("Test failed");
        }
    }
}

File: number theory and crypto/dart/prime.dart
  import 'dart:math';
  void main () {
    print(isPrime(2));
    print(isPrime(3));
    print(isPrime(9));
    List<int> primes = sieve(100);
    for (int i=0; i<primes.length; i++) {
      if (primes[i] == 1) print(i);
    }
  }

  bool isPrime(int n) {
    if (n<2) return false;
    else if (n == 2) return true;
    else if (n%2 == 0) return false;
    for (int i=3; i<=sqrt(n); i+=2) {
      if (n%i == 0) return false;
    }
    return true;
  }

  // sieve of eratosthenes
  List<int> sieve(int n) {
    // Declare a list of n+1 elements
    List<int> primes = List<int>.filled(n + 1, 1);

    // Initialize elements at index 0 and 1 to 0
    primes[0] = 0;
    primes[1] = 0;

    // Start from 2 and go up to n to find prime numbers
    for (int i = 2; i <= n; i++) {
      if (primes[i] == 1) {
        // Mark multiples of i as not prime
        for (int j = i * i; j <= n; j += i) {
          primes[j] = 0;
        }
      }
    }

    return primes;
  }


